torch/csrc/jit/passes/frozen_concat_linear.cpp

```diff
@@ -0,0 +1,186 @@
+          auto candidate_bias = constant_as<Tensor>(candidate_node->namedInput("bias"));
+          auto candidate_bias_type = candidate_bias->dtype();
+
+          // For now we will just keep it simple and require matching types
+          // Type promotion mgiht cause performance to actually decrease.
+          if (weight_type != candidate_weight_type || bias_type != candidate_bias_type) {          
+            it++;
+            continue;
+          }
+
+          bool can_move_before_all = true;
+          for (auto n : matching_nodes) {
+            can_move_before_all &=
+                getAliasDb()->couldMoveBeforeTopologically(candidate_node, n);
+          }
+          if (!can_move_before_all) {
+            it++;
+            continue;
+          }
+
+          // Found a node that is eligible for combination
+          auto prev = it;
+          it++;
+          uses.erase(prev);
+          matching_nodes.push_back(candidate_node);
+        }
+        if (matching_nodes.size() == 1) {
+          // Found no other compatible linear layers, do nothing.
+          continue;
+        }
+        graph_modified = true;
+
+        for (auto n : matching_nodes) {
+          if (n == base_node) {
+            continue;
+          }
+          getAliasDb()->moveBeforeTopologicallyValid(base_node, n);
+        }
+
+        // Now actually merge the candidate nodes
+        graph_->setInsertPoint(base_node);
+        Value* zero = graph_->insertConstant(0);
+
+        auto weight_list = c10::fmap(
+            matching_nodes, [](Node* n) { return n->namedInput("weight"); });
+        auto weight_with_dim(weight_list);
+        weight_with_dim.push_back(zero);
+
+        auto bias_with_dim = c10::fmap(
+            matching_nodes, [](Node* n) { return n->namedInput("bias"); });
+        bias_with_dim.push_back(zero);
+
+        auto weight_cat = graph_->create(prim::VarConcat, weight_with_dim);
+        weight_cat->insertBefore(base_node);
+        auto bias_cat = graph_->create(prim::VarConcat, bias_with_dim);
+        bias_cat->insertBefore(base_node);
+
+        // create the new node
+        std::vector<Value*> linear_in = {
+            op_value, weight_cat->output(), bias_cat->output()};
+        auto linear_node = graph_->create(aten::linear, linear_in);
+        linear_node->insertBefore(base_node);
+
+        // Update the outputs of the nodes
+        graph_->setInsertPoint(linear_node);
+        Value* neg1 = graph_->insertConstant(-1);
+        Value* one = graph_->insertConstant(1);
+
+        long long cur_loc = 0;
+        Value* cur_val = zero;
+
+        for (auto& orig_node : matching_nodes) {
+          Tensor weight_tensor = constant_as<Tensor>(orig_node->namedInput("weight")).value();
+          long long next_loc = cur_loc + weight_tensor.size(0);
+          Value* next_val = graph_->insertConstant(next_loc);
+            
+          auto slice = graph_->create(
```

Comment:
for any new Values we create, we would need to modify the AliasDb to maintain its correctness. I think for now I would either only accept linear layers which dont have modified inputs (this would not be a hard change), or regenerate the alias db on each transformation. We dont expect to do this vary frequently for any model so this might be fine but isn't the best....

if you imagine
```
x = linear(...)
y = linear(...)


x.add_(3)
```

after we combine x and y we would want the output of the first slice to still have its original aliasing properties. i think it would be simpler to just record which linear layer inputs / outputs are mutated, collect that in a list, and only optimize the layers that dont have mutated inputs. after we merge linear layers together we update the mutated list.

Reply:
Would it be fine if we processed the input tensor groups in reverse topological order (maybe that is too complicated)? Otherwise, I am happy to just go with the first option and keep it simple.

Reply:
Yea, i think reverse topological order would work as well

Reply:
[Note to self: Don't forget to add a test case where we have two interleaved optimizations needed so that I can test that this works correctly.]